src/plugins/intel_npu/src/plugin/npuw/util.cpp

```diff
@@ -827,6 +827,45 @@ ov::Tensor ov::npuw::util::concat(const std::vector<ov::Tensor>& tt, std::size_t
     }
 }
 
+void ov::npuw::util::permute_i4d(const ov::SoPtr<ov::ITensor>& src,
+                                 ov::SoPtr<ov::ITensor>& dst,
+                                 const std::array<int, 4> order) {
+    const auto& src_shape = src->get_shape();
+    const auto& dst_shape = dst->get_shape();
+    NPUW_ASSERT(src_shape.size() == 4);
+    NPUW_ASSERT(dst_shape.size() == 4);
+
+    const auto& src_s = src->get_strides();
+    const auto& dst_s = dst->get_strides();
+    const auto elem_size = src->get_byte_size() / src->get_size();
+
+    const auto* src_p = static_cast<uint8_t*>(src->data());
+    auto* dst_p = static_cast<uint8_t*>(dst->data());
+
+    for (std::size_t i = 0; i < src_shape[0]; i++) {
+        for (std::size_t j = 0; j < src_shape[1]; j++) {
+            for (std::size_t k = 0; k < src_shape[2]; k++) {
+                for (std::size_t l = 0; l < src_shape[3]; l++) {
+                    const std::size_t v_src[4] = {i, j, k, l};  // source vector
+                    const auto src_o =
+                        v_src[0] * src_s[0] + v_src[1] * src_s[1] + v_src[2] * src_s[2] + v_src[3] * src_s[3];
+
+                    const std::size_t v_dst[4] = {
+                        // for order 0,1,3,2:
+                        v_src[order[0]],  // i -> i
+                        v_src[order[1]],  // j -> j
+                        v_src[order[2]],  // k -> l
+                        v_src[order[3]],  // l -> k
+                    };
+                    const auto dst_o =
+                        v_dst[0] * dst_s[0] + v_dst[1] * dst_s[1] + v_dst[2] * dst_s[2] + v_dst[3] * dst_s[3];
+                    std::copy_n(src_p + src_o, elem_size, dst_p + dst_o);
```

Comment:
So based on my paper math this approach look correct.
This permute assumes that two dimensions swap their places (order), so given the position in the source tensor `(i/j/k/l)` we find the right position in the dst tensor with the remapped vector `(i/j/l/k)` in this case. As the axis order defines the physical layout, we use the src/dst strides in the same order as given. Finally, we copy `$(element_size)` bytes for a single element with `copy_n`.